system/jlib/jevent.hpp

```diff
@@ -0,0 +1,169 @@
+    void stopRecording();
+
+//Functions for each of the events that can be recorded..
+    void recordIndexLookup(unsigned fileid, offset_t offset, byte nodeKind, bool hit);
+    void recordIndexLoad(unsigned fileid, offset_t offset, byte nodeKind, size32_t size, __uint64 elapsedTime, __uint64 readTime);
+    void recordIndexEviction(unsigned fileid, offset_t offset, byte nodeKind, size32_t size);
+
+    void recordDaliConnect(const char * path, __uint64 id);
+    void recordDaliDisconnect(__uint64 id);
+
+    //-------------------------- End of the public interface --------------------------
+
+protected:
+    void checkAttrValue(EventAttr attr, size_t size);
+
+    void writeEventHeader(EventType type, offset_type & offset);
+    void writeEventFooter(offset_type & offset, size32_t requiredSize, offset_t writeOffset);
+
+    template <class T>
+    void write(offset_type & offset, T value)
+    {
+        writeData(offset, sizeof(value), &value);
+    }
+
+    void write(offset_type & offset, const char * value)
+    {
+        writeData(offset, strlen(value)+1, value);
+    }
+
+    template <class T>
+    void write(EventRecorder::offset_type & offset, EventAttr attr, T value)
+    {
+        checkAttrValue(attr, sizeof(value));        // MORE: make this debug only....
+        writeData(offset, sizeof(attr), &attr);
+        writeData(offset, sizeof(value), &value);
+    }
+
+    void write(EventRecorder::offset_type & offset, EventAttr attr, const char * value)
+    {
+        writeData(offset, sizeof(attr), &attr);
+        writeData(offset, strlen(value)+1, value);
+    }
+
+    void writeData(offset_type & offset, size_t size, const void * data);
+
+    offset_type reserveEvent(size32_t size);
+    void commitEvent(offset_type startOffset, size32_t size);
+
+    void writeBlock(offset_type startOffset, size32_t size);
+
+    unsigned getBlockFromOffset(offset_type offset) { return (unsigned)((offset & bufferMask) / OutputBlockSize); }
+
+protected:
+    static constexpr size32_t numBlocks = 2;
+    static constexpr size32_t OutputBlockSize = 0x100000;
+    static constexpr size32_t OutputBufferSize = numBlocks * OutputBlockSize;
+    static constexpr offset_t bufferMask = OutputBufferSize-1;
+    static constexpr offset_t blockMask = OutputBlockSize-1;
+
+    std::atomic<bool> recordingEvents{false};
+    offset_type nextOffset{0};
+    offset_type nextWriteOffset{0};
+    unsigned counts[numBlocks] = {0};
+    cycle_t startCycles{0};
+    MemoryAttr buffer;
+    CriticalSection cs;
+    unsigned sizeMessageHeaderFooter{0};
+    unsigned options{0};
+    Owned<IFileIO> output;
+};
+
+// The implementation exposes a global object so that the test for whether events are being recorded
+// is as lightweight as possible.
+// All access should be via the non-internal helper accessor functions
+
+namespace EventRecorderInternal
+{
+extern jlib_decl EventRecorder eventRecorder;
```

Comment:
I assume this is supposed to be a reference.

Reply:
no.  It is a external variable declaration.